File: examples/node/nestjs/src/chart/chart.controller.ts
import { Controller, Post, Body, Get, HttpException, HttpStatus } from '@nestjs/common';
import { ChartService } from './chart.service';
import type { ChartRequest, ChartResponse, ErrorResponse } from '../types';

@Controller('api/chart')
export class ChartController {
  constructor(private readonly chartService: ChartService) {}

  @Post()
  async generateChart(@Body() body: ChartRequest): Promise<ChartResponse> {
    const { prompt, context } = body;

    if (!prompt || typeof prompt !== 'string') {
      const error: ErrorResponse = {
        error: 'Missing or invalid prompt',
        code: 'INVALID_REQUEST',
      };
      throw new HttpException(error, HttpStatus.BAD_REQUEST);
    }

    try {
      return await this.chartService.resolve(prompt, context);
    } catch (err) {
      console.error('Chart generation error:', err);

      const isInvalidIntent = err instanceof Error && err.message.includes('Invalid intent');
      const error: ErrorResponse = {
        error: err instanceof Error ? err.message : 'Internal server error',
        code: isInvalidIntent ? 'INVALID_INTENT' : 'INTERNAL_ERROR',
      };

      throw new HttpException(error, isInvalidIntent ? HttpStatus.BAD_REQUEST : HttpStatus.INTERNAL_SERVER_ERROR);
    }
  }
}

@Controller('health')
export class HealthController {
  @Get()
  check() {
    return { status: 'ok', timestamp: new Date().toISOString() };
  }
}

File: examples/java/springboot/src/main/java/com/promptchart/controller/ChartController.java
package com.promptchart.controller;

import com.promptchart.model.ChartRequest;
import com.promptchart.model.ChartResponse;
import com.promptchart.service.IntentResolverService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.Instant;
import java.util.Map;

@RestController
@CrossOrigin(origins = "*")
public class ChartController {

    private final IntentResolverService resolverService;

    public ChartController(IntentResolverService resolverService) {
        this.resolverService = resolverService;
    }

    @PostMapping("/api/chart")
    public ResponseEntity<?> generateChart(@RequestBody ChartRequest request) {
        if (request.prompt() == null || request.prompt().isBlank()) {
            return ResponseEntity.badRequest()
                .body(Map.of("error", "Missing or invalid prompt", "code", "INVALID_REQUEST"));
        }

        try {
            ChartResponse response = resolverService.resolve(request.prompt());
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            return ResponseEntity.internalServerError()
                .body(Map.of("error", e.getMessage(), "code", "INTERNAL_ERROR"));
        }
    }

    @GetMapping("/health")
    public ResponseEntity<?> health() {
        return ResponseEntity.ok(Map.of("status", "ok", "timestamp", Instant.now().toString()));
    }
}
